x/virtualgroup/keeper/msg_server.go

```diff
@@ -443,6 +444,26 @@ func (k msgServer) StorageProviderExit(goCtx context.Context, msg *types.MsgStor
 		return nil, sptypes.ErrStorageProviderExitFailed.Wrapf("sp not in service, status: %s", sp.Status.String())
 	}
 
+	stat, found := k.GetGVGStatisticsWithinSP(ctx, sp.Id)
+	if found && stat.BreakRedundancyReqmtGvgCount != 0 {
+		return nil, types.ErrSPCanNotExit.Wrapf("The SP has %d GVG that break the redundancy requirement, need to be resolved before exit.", stat.BreakRedundancyReqmtGvgCount)
+	}
+
+	// can only allow 1 sp exit at a time, a GVG can have only 1 SwapInInfo associated.
+	exittingSPNum := uint32(0)
+	sps := k.spKeeper.GetAllStorageProviders(ctx)
+	for _, curSP := range sps {
+		// TODO Remove FORCE_EXIT SP info from chain
+		if curSP.Status == sptypes.STATUS_GRACEFUL_EXITING ||
+			curSP.Status == sptypes.STATUS_FORCE_EXITING {
+			exittingSPNum++
+		}
+	}
+	maxSPExitingNum := k.SpConcurrentExitNum(ctx)
```

Comment:
will be renamed

Reply:
Actually the testnet can not run from genesis already, because of the  `gas read` issue on testnet, @forcodedancing  know this well, so I am thinking do not use hardfor for the logic change in these handler?

Reply:
By the way, can we compile this branch and run a fullnode on testet to verify whether it  break anything?

Reply:
The mainnet does not send any SP exit like tx, so it is safe. For most testnet fullnode, they can still use state sync to catch up.